File: app/src/main/java/com/samurai/morseencoder/MainActivityTabPagerAdapter.java
package com.samurai.morseencoder;

/**
 * Created by Sergey on 24.01.2017.
 */

import android.content.Context;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;

import com.samurai.morseencoder.fragments.MainFrame;
import com.samurai.morseencoder.fragments.Rules;
import com.samurai.morseencoder.fragments.Settings;

public class MainActivityTabPagerAdapter extends FragmentPagerAdapter {

    private Context context;
    private final int PAGE_COUNT=3;

    public MainActivityTabPagerAdapter(FragmentManager fm , Context context) {
        super(fm);
        this.context = context;
    }

    @Override
    public Fragment getItem(int arg0) {
        switch (arg0) {
            case 0:
                return new MainFrame();
            case 1:
                return new Settings();
            case 2:
                return new Rules();
            default:
                break;
        }
        return null;
    }

    @Override
    public int getCount() {
        return PAGE_COUNT;
    }

    @Override public CharSequence getPageTitle(int position) {
        String[] tabTitles = context.getResources().getStringArray(R.array.tab_names);
        return tabTitles[position];
    }
}

File: app/src/main/kotlin/com/samurai/morseencoder/MainActivityTabPagerAdapter.kt
package com.samurai.morseencoder

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.samurai.morseencoder.fragments.TranslationFragment
import com.samurai.morseencoder.fragments.translation_rules.TranslationRulesFragment
import com.samurai.morseencoder.fragments.SettingsFragment
import com.samurai.sysequsol.R

class MainActivityTabPagerAdapter(activity: FragmentActivity?) : FragmentStateAdapter(activity!!) {

    private val mFragmentList: MutableList<Fragment> = mutableListOf(
        TranslationFragment(), SettingsFragment(), TranslationRulesFragment()
    )

    private var mFragmentTitleList: MutableList<String> = ArrayList()

    init {
        mFragmentTitleList = activity?.resources?.getStringArray(R.array.tab_names)
            ?.toMutableList()!!
    }

    fun getTabTitle(position: Int): String {
        return mFragmentTitleList[position]
    }

    fun addFragment(fragment: Fragment, title: String) {
        mFragmentList.add(fragment)
        mFragmentTitleList.add(title)
    }

    override fun getItemCount(): Int {
        return mFragmentList.size
    }

    override fun createFragment(position: Int): Fragment {
        return mFragmentList[position]
    }
}